vectorizers/timed_token_cooccurrence_vectorizer.py

```diff
@@ -0,0 +1,481 @@
+
+    Note: The algorithm assumes the matrix is an hstack of cooccurrence matrices with the same vocabulary,
+    with kernel and window parameters given in the same order.
+
+    Parameters
+    ----------
+
+    token_sequences: Iterable of Iterables
+        The collection of  (token, time_stamp)  sequences to generate skip-gram data for.
+
+    window_size_array : numpy.ndarray of shape(n, n_vocab)
+        The collection of window sizes per token per directed cooccurrence
+
+    window_reversals: numpy.array(bool)
+        The collection of indicators whether or not the window is after the target token.
+
+    kernel_functions: tuple
+        The n-tuple of kernel functions
+
+    kernel_args: tuple(tuples)
+        The n-tuple of update_kernel args per kernel function
+
+    mix_weights: tuple
+        The n-tuple of mix weights to apply to the kernel functions
+
+    n_unique_tokens: int
+        The number of unique tokens
+
+    prior_indices:  numpy.array
+        The csr indices of the hstacked cooccurrence matrix
+
+    prior_indptr: numpy.array
+        The csr indptr of the hstacked cooccurrence matrix
+
+    prior_data: numpy.array
+        The csr data of the hstacked cooccurrence matrix
+
+    Returns
+    -------
+    posterior_data: numpy.array
+        The data of the updated csr matrix after one iteration of EM.
+
+    """
+
+    posterior_data = np.zeros_like(prior_data)
+    n_windows = window_size_array.shape[0]
+    window_reversal_const = np.zeros(len(window_reversals)).astype(np.int32)
+    window_reversal_const[window_reversals] = 1
+
+    for d_i, seq in enumerate(token_sequences):
+        for w_i, target_pair in enumerate(seq):
+            windows = []
+            kernels = []
+            target_word = np.int32(target_pair[0])
+            target_time = target_pair[1]
+            for i in range(n_windows):
+                win = window_at_index(
+                    seq,
+                    window_size_array[i, target_word],
+                    w_i,
+                    reverse=window_reversals[i],
+                )
+
+                this_window = np.array([w[0] for w in win], dtype=np.int32)
+                time_deltas = np.array([np.abs(w[1] - target_time) for w in win])
+                this_kernel = mix_weights[i] * kernel_functions[i](
+                    this_window, time_deltas, *kernel_args[i]
+                )
+                windows.append(this_window)
+                kernels.append(this_kernel)
+
+            posterior_data = em_update_matrix(
+                posterior_data,
+                prior_indices,
+                prior_indptr,
+                prior_data,
+                n_unique_tokens,
+                target_word,
+                windows,
+                kernels,
+            )
+
+    return posterior_data
+
+
+class TimedTokenCooccurrenceVectorizer(BaseCooccurrenceVectorizer):
+    """Given a sequence, or list of sequences of tokens, produce a collection of directed
```

Comment:
We should probably include a few details about the (token, time) pair, that this class is designed to tackle, in the docstring